File: research/experiments/sweep/runs/opt-large-csharp-sonnet45/Optimization/ResultTypes.cs
namespace Optimization;

/// <summary>
/// Options for optimization algorithms.
/// </summary>
public class OptimizeOptions
{
    public double GradTol { get; set; } = 1e-8;
    public double StepTol { get; set; } = 1e-8;
    public double FuncTol { get; set; } = 1e-12;
    public int MaxIterations { get; set; } = 1000;
}

/// <summary>
/// Result of an optimization run.
/// </summary>
public class OptimizeResult
{
    public double[] X { get; set; } = Array.Empty<double>();
    public double Fun { get; set; }
    public double[]? Gradient { get; set; }
    public int Iterations { get; set; }
    public int FunctionCalls { get; set; }
    public int GradientCalls { get; set; }
    public bool Converged { get; set; }
    public string Message { get; set; } = "";
}

/// <summary>
/// Reason for convergence or termination.
/// </summary>
public abstract class ConvergenceReason
{
    public abstract string Kind { get; }
}

public class GradientConvergence : ConvergenceReason
{
    public override string Kind => "gradient";
}

public class StepConvergence : ConvergenceReason
{
    public override string Kind => "step";
}

public class FunctionConvergence : ConvergenceReason
{
    public override string Kind => "function";
}

public class MaxIterationsReached : ConvergenceReason
{
    public override string Kind => "maxIterations";
}

public class LineSearchFailure : ConvergenceReason
{
    public override string Kind => "lineSearchFailed";
}

/// <summary>
/// Shared convergence logic for optimization algorithms.
/// </summary>
public static class Convergence
{
    /// <summary>Create default options with optional overrides.</summary>
    public static OptimizeOptions DefaultOptions(OptimizeOptions? overrides = null)
    {
        if (overrides == null) return new OptimizeOptions();

        return new OptimizeOptions
        {
            GradTol = overrides.GradTol,
            StepTol = overrides.StepTol,
            FuncTol = overrides.FuncTol,
            MaxIterations = overrides.MaxIterations
        };
    }

    /// <summary>
    /// Check convergence criteria in order: gradient → step → function → maxIterations.
    /// Returns the first satisfied criterion, or null if none are met.
    /// </summary>
    public static ConvergenceReason? CheckConvergence(
        double gradNorm,
        double stepNorm,
        double funcChange,
        int iteration,
        OptimizeOptions opts)
    {
        if (gradNorm < opts.GradTol)
            return new GradientConvergence();

        if (stepNorm < opts.StepTol)
            return new StepConvergence();

        if (funcChange < opts.FuncTol)
            return new FunctionConvergence();

        if (iteration >= opts.MaxIterations)
            return new MaxIterationsReached();

        return null;
    }

    /// <summary>
    /// Returns true for gradient/step/function convergence,
    /// false for maxIterations/lineSearchFailed.
    /// </summary>
    public static bool IsConverged(ConvergenceReason reason)
    {
        return reason.Kind is "gradient" or "step" or "function";
    }

    /// <summary>Human-readable convergence message.</summary>
    public static string ConvergenceMessage(ConvergenceReason reason)
    {
        return reason.Kind switch
        {
            "gradient" => "Converged: gradient norm below tolerance",
            "step" => "Converged: step size below tolerance",
            "function" => "Converged: function change below tolerance",
            "maxIterations" => "Terminated: maximum iterations reached",
            "lineSearchFailed" => "Terminated: line search failed",
            _ => "Unknown termination reason"
        };
    }
}

File: research/experiments/sweep/runs/opt-medium-go-sonnet45/result_types.go
package optimization

// OptimizeOptions contains convergence tolerances and iteration limits.
type OptimizeOptions struct {
	GradTol        float64
	StepTol        float64
	FuncTol        float64
	MaxIterations  int
}

// OptimizeResult contains the result of an optimization run.
type OptimizeResult struct {
	X             []float64
	Fun           float64
	Gradient      []float64 // nil for derivative-free methods
	Iterations    int
	FunctionCalls int
	GradientCalls int
	Converged     bool
	Message       string
}

// ConvergenceReason describes why an optimization terminated.
type ConvergenceReason struct {
	Kind string
}

// DefaultOptions returns default optimization options with optional overrides.
func DefaultOptions(overrides *OptimizeOptions) OptimizeOptions {
	opts := OptimizeOptions{
		GradTol:       1e-8,
		StepTol:       1e-8,
		FuncTol:       1e-12,
		MaxIterations: 1000,
	}

	if overrides != nil {
		if overrides.GradTol != 0 {
			opts.GradTol = overrides.GradTol
		}
		if overrides.StepTol != 0 {
			opts.StepTol = overrides.StepTol
		}
		if overrides.FuncTol != 0 {
			opts.FuncTol = overrides.FuncTol
		}
		if overrides.MaxIterations != 0 {
			opts.MaxIterations = overrides.MaxIterations
		}
	}

	return opts
}

// CheckConvergence checks convergence criteria in order: gradient → step → function → maxIterations.
// Returns nil if no criterion is met.
func CheckConvergence(gradNorm, stepNorm, funcChange float64, iteration int, opts OptimizeOptions) *ConvergenceReason {
	if gradNorm < opts.GradTol {
		return &ConvergenceReason{Kind: "gradient"}
	}
	if stepNorm < opts.StepTol {
		return &ConvergenceReason{Kind: "step"}
	}
	if funcChange < opts.FuncTol {
		return &ConvergenceReason{Kind: "function"}
	}
	if iteration >= opts.MaxIterations {
		return &ConvergenceReason{Kind: "maxIterations"}
	}
	return nil
}

// IsConverged returns true if the convergence reason indicates successful convergence.
func IsConverged(reason *ConvergenceReason) bool {
	if reason == nil {
		return false
	}
	switch reason.Kind {
	case "gradient", "step", "function":
		return true
	case "maxIterations", "lineSearchFailed":
		return false
	default:
		return false
	}
}

// ConvergenceMessage returns a human-readable message for the convergence reason.
func ConvergenceMessage(reason *ConvergenceReason) string {
	if reason == nil {
		return "Not converged"
	}
	switch reason.Kind {
	case "gradient":
		return "Converged: gradient norm below tolerance"
	case "step":
		return "Converged: step size below tolerance"
	case "function":
		return "Converged: function change below tolerance"
	case "maxIterations":
		return "Terminated: maximum iterations reached"
	case "lineSearchFailed":
		return "Terminated: line search failed"
	default:
		return "Unknown termination reason"
	}
}
